sklearn/multiclass.py

```diff
@@ -1189,34 +1265,63 @@ def fit(self, X, y, **fit_params):
-        check_classification_targets(y)
+        # we postpone the validation of X to the `fit` of the underlying estimators but
+        # we need this check for the validation of `code_size`.
+        n_samples = _num_samples(X)
+        if n_samples < 1:
+            raise ValueError(
+                f"Found array {n_samples} (shape={X.shape}) while minimum of 1 is "
+                "required."
+            )
 
-        self.classes_ = np.unique(y)
-        n_classes = self.classes_.shape[0]
-        if n_classes == 0:
+        if self.decoding == "cityblock" and not hasattr(
+            self.estimator, "predict_proba"
+        ):
+            raise ValueError(
+                "The estimator does not have a `predict_proba` method. "
+                "Thus, the 'cityblock' decoding strategy is not supported."
+            )
+        elif self.decoding == "loss" and not (
+            hasattr(self.estimator, "predict_proba")
+            or hasattr(self.estimator, "decision_function")
+        ):
             raise ValueError(
-                "OutputCodeClassifier can not be fit when no class is present."
+                "The estimator does not have either a `predict_proba` or "
+                "`decision_function` method. Thus, the 'loss' decoding strategy is "
+                "not supported."
             )
-        n_estimators = int(n_classes * self.code_size)
 
-        # FIXME: there are more elaborate methods than generating the codebook
-        # randomly.
-        self.code_book_ = random_state.uniform(size=(n_classes, n_estimators))
-        self.code_book_[self.code_book_ > 0.5] = 1.0
+        random_state = check_random_state(self.random_state)
+        check_classification_targets(y)
 
-        if hasattr(self.estimator, "decision_function"):
-            self.code_book_[self.code_book_ != 1] = -1.0
-        else:
-            self.code_book_[self.code_book_ != 1] = 0.0
+        self._label_encoder = LabelEncoder().fit(y)
+        y_encoded = self._label_encoder.transform(y)
+        n_classes = len(self._label_encoder.classes_)
+        code_size = int(n_classes * self.code_size)
+        n_binary_values = 2**code_size
 
-        classes_index = {c: i for i, c in enumerate(self.classes_)}
+        if n_binary_values < n_classes:
+            raise ValueError(
+                "The code book size is not big enough to encode all classes. Thus, "
+                "different classes will share the same code. Increase `code_size`. The "
+                "minimum value for `code_size` is"
+                f" {np.log2(n_classes) / n_classes:.2f}",
+            )
 
-        Y = np.array(
-            [self.code_book_[classes_index[y[i]]] for i in range(_num_samples(y))],
-            dtype=int,
+        # TODO: when supporting random generator from NumPy (i.e.
+        # `np.random.default_rng()`), the following can use `random_state.choice`
+        # instead of the `random` Python build-in module.
```

Comment:
why not get a seed, like you're getting here, and then get a numpy generator, especially if the numpy of the user has it, so that we don't do behavior change later.